man/man3/Cjt_clusters.3
.TH "Cjt_clusters" 3 "Dc Mai 20 2020" "Version v6.3 19/05/2020" "Creació d'un arbre filogenètic. Xavier Coll Ribas" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Cjt_clusters \- Representa un conjunt de clústers\&.  

.SH SYNOPSIS
.br
.PP
.SS "Mètodes públics"

.in +1c
.ti -1c
.RI "\fBCjt_clusters\fP ()"
.br
.RI "Crea un conjunt de clústers buit\&. "
.ti -1c
.RI "double \fBdist_clust\fP (const string &id, const string &id2) const"
.br
.RI "Busca la distància entre els clústers demanats\&. "
.ti -1c
.RI "pair< string, string > \fBmin_dist\fP () const"
.br
.RI "Calcula la distància mínima entre clústers\&. "
.ti -1c
.RI "void \fBinicialitza_clusters\fP (\fBCjt_especies\fP &conjunt, const bool ve_de_func)"
.br
.RI "Inicialitza els clústers del conjunt\&. "
.ti -1c
.RI "void \fBejecuta_paso_wpgma\fP (const bool &a)"
.br
.RI "Fusiona els clústers que estan a menor distància\&. "
.ti -1c
.RI "void \fBimprime_cluster\fP (string id) const"
.br
.RI "Imprimeix el clúster indicat\&. "
.ti -1c
.RI "void \fBimprime_tabla_distancias\fP () const"
.br
.RI "Imprimeix la taula de distàncies entre clústers\&. "
.ti -1c
.RI "void \fBimprime_arbol_filogenetico\fP (\fBCjt_especies\fP &conjunt)"
.br
.RI "Imprimeix l'arbre filogenètic dels clústers resultants\&. "
.in -1c
.SS "Mètodes Privats"

.in +1c
.ti -1c
.RI "void \fBtabla_dist_clust\fP (const \fBCjt_especies\fP &conjunt)"
.br
.RI "Passa les distàncies entre espècies a la taula de distàncies entre clústers\&. "
.ti -1c
.RI "void \fBafegeix_especie_clusters\fP (const pair< string, string > &dist)"
.br
.RI "Afegeix la fusió dels dos clústers al conjunt de clústers i a la taula de distàncies\&. "
.ti -1c
.RI "void \fBelimina_especie_clusters\fP (const string &id)"
.br
.RI "Elimina la especie indicada del conjunt de clústers i de la taula de distàncies\&. "
.in -1c
.SS "Atributs Privats"

.in +1c
.ti -1c
.RI "map< string, \fBCluster\fP > \fBmap_clusters\fP"
.br
.RI "Estructura de dades on es guarden els clústers resultants\&. Conté un identificador i un clúster\&. "
.ti -1c
.RI "map< string, map< string, double > > \fBtabla_distancias_cluster\fP"
.br
.RI "Estructura de dades on guardem les distancies entre clústers\&. "
.in -1c
.SH "Descripció Detallada"
.PP 
Representa un conjunt de clústers\&. 

Ofereix la possibilitat d'ajuntar clústers per tal d'anar creant l'arbre filogenètic\&. També pots demanar imprimir un clúster determinat si existeix i imprimir l'arbre resultant\&. També ofereix una taula de distancies entre clústers\&. Els clústers es generen mitjançant la sincronització amb un conjunt d'especies i es compleix amb la funció inicializa_clústers\&.
.PP
\fBInvariant\fP
.RS 4
El conjunt de clústers tindrà un identificador que serà el mateix que el d'una especie i sempre estarà ordenat alfabèticament, aquest no té mida màxima ni mínima\&. La taula de distànciesentre clústers té els mateixos identificadors que els d'especie i també està ordenada alfabèticament\&. 
.RE
.PP

.PP
Definició a la línia 36 del fitxer Cjt_clusters\&.hh\&.
.SH "Documentació del Constructor i el Destructor"
.PP 
.SS "Cjt_clusters::Cjt_clusters ()"

.PP
Crea un conjunt de clústers buit\&. 
.PP
\fBPrecondició\fP
.RS 4
Cert\&. 
.RE
.PP
\fBPostcondició\fP
.RS 4
El resultat és un conjunt de clústers buit\&. 
.RE
.PP

.PP
Definició a la línia 19 del fitxer Cjt_clusters\&.cc\&.
.PP
.nf
19                            {
20     //Es crea el mapa de clústers i el de distancies buit\&.
21     map_clusters = map<string,Cluster>();
22     tabla_distancias_cluster = map<string,map<string,double>>();
23 }
.fi
.SH "Documentació de les Funcions Membre"
.PP 
.SS "void Cjt_clusters::tabla_dist_clust (const \fBCjt_especies\fP & conjunt)\fC [private]\fP"

.PP
Passa les distàncies entre espècies a la taula de distàncies entre clústers\&. 
.PP
\fBPrecondició\fP
.RS 4
Cert\&. 
.RE
.PP
\fBPostcondició\fP
.RS 4
S'ha passat totes les distancies entre especies al conjunt de clústers\&. 
.RE
.PP
\fBParàmetres\fP
.RS 4
\fI\fBCjt_especies\fP\fP Entra un objecte del tipus conjunt especies\&. 
.RE
.PP

.PP
Definició a la línia 146 del fitxer Cjt_clusters\&.cc\&.
.PP
.nf
146                                                                {
147     //Consulta mitjançant el conjunt de espècies les distancies entre elles per 
148     //passar-les al conjunt de clústers\&.
149     for (map<string,Cluster>::const_iterator it = map_clusters\&.begin(); it != map_clusters\&.end(); ++it) {
150         map<string, double> aux;
151         for (map<string,Cluster>::const_iterator it2 = it; it2 != map_clusters\&.end(); ++it2) {
152             if (it->first != it2->first) { //Comprova que el iterador sigui diferent al segon iterador\&.
153             double x = conjunt\&.calcular_distancia(it->first, it2->first);
154             aux\&.insert(make_pair(it2->first, x));
155             }
156         }
157         tabla_distancias_cluster\&.insert(make_pair(it->first, aux));
158     }
159 }
.fi
.SS "void Cjt_clusters::afegeix_especie_clusters (const pair< string, string > & dist)\fC [private]\fP"

.PP
Afegeix la fusió dels dos clústers al conjunt de clústers i a la taula de distàncies\&. 
.PP
\fBPrecondició\fP
.RS 4
El pair no és buit\&. 
.br
 
.RE
.PP
\fBPostcondició\fP
.RS 4
S'ha fusionat els clústers\&. 
.RE
.PP

.PP
Definició a la línia 162 del fitxer Cjt_clusters\&.cc\&.
.PP
.nf
162                                                                            {
163     //Aquí es modifica les distancies dels clústers fent servir
164     //la formula de WPGMA\&.
165     string fus = dist\&.first + dist\&.second;
166     map<string,map<string,double>>::const_iterator it = tabla_distancias_cluster\&.begin();
167     while (it != tabla_distancias_cluster\&.end() and it->first < fus) {
168         tabla_distancias_cluster[it->first][fus] = (tabla_distancias_cluster[it->first][dist\&.first] + tabla_distancias_cluster[min(it->first,dist\&.second)][max(it->first,dist\&.second)])/2;
169         ++it;
170     }
171     while (it != tabla_distancias_cluster\&.end()) {
172         tabla_distancias_cluster[fus][it->first] = (tabla_distancias_cluster[dist\&.first][it->first] + tabla_distancias_cluster[min(it->first,dist\&.second)][max(it->first,dist\&.second)])/2;
173         ++it;
174     }
175     map<string,Cluster>::const_iterator a = map_clusters\&.find(dist\&.first);
176     map<string,Cluster>::const_iterator b = map_clusters\&.find(dist\&.second);
177     Cluster c(a->second,b->second,dist_clust(dist\&.first,dist\&.second));
178     map_clusters\&.insert(make_pair(fus,c));
179 
180 
181 }
.fi
.SS "void Cjt_clusters::elimina_especie_clusters (const string & id)\fC [private]\fP"

.PP
Elimina la especie indicada del conjunt de clústers i de la taula de distàncies\&. 
.PP
\fBPrecondició\fP
.RS 4
El clúster a eliminar existeix\&. 
.br
 
.RE
.PP
\fBPostcondició\fP
.RS 4
S'ha eliminat el clúster del conjunt de clústers i de la taula de distancies\&. 
.RE
.PP

.PP
Definició a la línia 185 del fitxer Cjt_clusters\&.cc\&.
.PP
.nf
185                                                             {
186     //Es posiciona un iterador apuntant a la posició de l'id demanat\&. El nom de l'iterador era per donar humor al còdig\&.
187     map<string, Cluster>::const_iterator it_peix = map_clusters\&.find(id);
188   map_clusters\&.erase(it_peix);
189     //Es posiciona un iterador apuntant a la posició de l'id demanat\&. 
190   map<string,map<string,double>>::const_iterator it = tabla_distancias_cluster\&.find(id);
191     //Es posiciona un iterador apuntant a la primera posició del mapa de distancies\&.
192   //Inv: Farà tantes iteracions fins que l'iterador sigui igual al primer que hem buscat\&. Així ens
193   //estalviem bucles innecessaris\&. Aquest bucle elimina la espècie demanada del map interior\&.
194   for (map<string,map<string,double>>::iterator it2 = tabla_distancias_cluster\&.begin(); it2 != it; ++it2) {
195       it2->second\&.erase(it2->second\&.find(id));
196     }
197   //Eliminem la espècie del mapa gran de la taula de distancies\&.
198   tabla_distancias_cluster\&.erase(it);
199 
200 }
.fi
.SS "double Cjt_clusters::dist_clust (const string & id, const string & id2) const"

.PP
Busca la distància entre els clústers demanats\&. 
.PP
\fBPrecondició\fP
.RS 4
Els dos clústers demanats existeixen\&. 
.RE
.PP
\fBPostcondició\fP
.RS 4
S'ha retornat la distància entre els clústers\&. 
.RE
.PP
\fBParàmetres\fP
.RS 4
\fIstring\fP Entren per referència constant dos string amb els identificadors dels clústers\&. 
.RE
.PP
\fBRetorna\fP
.RS 4
Retorna un double amb la distància que els separa genèticament\&. Retorna -1 si algun dels dos clústers no existeix\&. 
.RE
.PP

.PP
Definició a la línia 29 del fitxer Cjt_clusters\&.cc\&.
.PP
.nf
29                                                                         {
30     //Funció per consultar les distancies dins el mapa de 
31     //distancies de clústers\&.
32         map<string,map<string,double>>::const_iterator it = tabla_distancias_cluster\&.find(id);
33       map<string,double>::const_iterator it2 = it->second\&.find(id2);
34       return it2->second;
35 }
.fi
.SS "pair< string, string > Cjt_clusters::min_dist () const"

.PP
Calcula la distància mínima entre clústers\&. 
.PP
\fBPrecondició\fP
.RS 4
Cert\&. 
.RE
.PP
\fBPostcondició\fP
.RS 4
S'ha calculat la distancia mínima entre els clústers del conjunt\&. 
.RE
.PP
\fBRetorna\fP
.RS 4
Retorna un pair\&. Els dos elements són de tipu string, retorna els identificadors de clústers a menor distancia\&. 
.RE
.PP

.PP
Definició a la línia 38 del fitxer Cjt_clusters\&.cc\&.
.PP
.nf
38                                                  {
39     //Es busca els clústers a menor distancia i 
40     //es retorna en una pair els dos clústers\&.
41     double distancia = 101;
42     pair<string,string> dists;
43     for (map<string,map<string,double>>::const_iterator it = tabla_distancias_cluster\&.begin(); it != tabla_distancias_cluster\&.end(); ++it) {
44         for (map<string,double>::const_iterator it2 = it->second\&.begin(); it2 != it->second\&.end(); ++it2) {
45                 if (it2->second < distancia ){
46                 distancia = it2->second;
47                 dists\&.first = it->first;
48                 dists\&.second = it2->first;
49                 }
50             }
51 
52         }
53     return dists;
54 
55 }
.fi
.SS "void Cjt_clusters::inicialitza_clusters (\fBCjt_especies\fP & conjunt, const bool ve_de_func)"

.PP
Inicialitza els clústers del conjunt\&. 
.PP
\fBPrecondició\fP
.RS 4
Existeix un conjunt d'especies\&. 
.RE
.PP
\fBPostcondició\fP
.RS 4
Els clústers s'han inicialitzat i estan dins del conjunt de clústers ordenats alfabèticament\&. 
.RE
.PP
\fBParàmetres\fP
.RS 4
\fI\fBCjt_especies\fP\fP Entra un conjunt d'especies per referència\&. 
.br
\fIBool\fP El booleà té una gran importància, si bool = true imprimirà la taula de distàncies entre clústers, en cas que bool = fals, la funció no imprimirà res\&. 
.RE
.PP

.PP
Definició a la línia 60 del fitxer Cjt_clusters\&.cc\&.
.PP
.nf
60                                                                                    {
61     //Establim els iteradors del conjunt d'espècies per tal de poder fer el bucle
62     //A cada iteració es modifica una string, agafant la id de una espècie i es
63     //crea un clúster a partir de la string i la introdueix al mapa de clústers\&.
64     map_clusters\&.clear();
65     tabla_distancias_cluster\&.clear();
66     conjunt\&.inicio(); // Iterador del conjunt d'espècies que apunta a la primera posició\&.
67          while (not conjunt\&.final()) { //comprova que l'iterador no apunta al final
68              string x;
69              conjunt\&.actual(x); // La string x agafa el valor del iterador\&.
70              Cluster clust(x);
71              map_clusters\&.insert(make_pair(x, clust));
72              conjunt\&.avanza(); //Avança posició del iterador\&.
73 
74         }
75     //Es crea la taula de distancies a partir del conjunt\&.
76     tabla_dist_clust(conjunt);
77     if (ve_de_func) imprime_tabla_distancias();
78 }
.fi
.SS "void Cjt_clusters::ejecuta_paso_wpgma (const bool & a)"

.PP
Fusiona els clústers que estan a menor distància\&. 
.PP
\fBPrecondició\fP
.RS 4
La mida del conjunt de clústers és més gran que 1\&. 
.RE
.PP
\fBPostcondició\fP
.RS 4
S'ha fusionat els clústers i s'ha actualitzat la taula de distàncies\&. 
.RE
.PP
\fBParàmetres\fP
.RS 4
\fIbool\&.\fP Si el bool = true, s'imprimeix la taula de distancies entre clústers resultant\&. 
.RE
.PP

.PP
Definició a la línia 81 del fitxer Cjt_clusters\&.cc\&.
.PP
.nf
81                                                    {
82     //S'executa el pas wpgma sempre i quan el tamany
83     //del mapa de clústers sigui > 1\&.
84     if (map_clusters\&.size() > 1) {
85     pair<string,string> pair_dist = min_dist(); //pair amb els dos clústers a menor distancia\&.
86     afegeix_especie_clusters(pair_dist); //Es fica els dos clústers i es fusionen en un nou clúster\&.
87     elimina_especie_clusters(pair_dist\&.first); //S'elimina el clúster\&.
88     elimina_especie_clusters(pair_dist\&.second); //S'elimina el clúster\&.
89     if (a) imprime_tabla_distancias();
90     }
91     else cout << "ERROR: num_clusters <= 1" << endl;
92 }
.fi
.SS "void Cjt_clusters::imprime_cluster (string id) const"

.PP
Imprimeix el clúster indicat\&. 
.PP
\fBPrecondició\fP
.RS 4
Existeix el clúster indicat\&. 
.RE
.PP
\fBPostcondició\fP
.RS 4
S'ha imprès el clúster indicat\&. 
.RE
.PP
\fBParàmetres\fP
.RS 4
\fIstring\fP id 
.RE
.PP

.PP
Definició a la línia 98 del fitxer Cjt_clusters\&.cc\&.
.PP
.nf
98                                                   {
99     //Es posa un iterador apuntant al clúster indicat
100     //Si no apunta al final, s'imprimeix, sino surt un missatge d'error\&.
101     map<string,Cluster>::const_iterator it = map_clusters\&.find(id);
102     if (it == map_clusters\&.end()) cout << "ERROR: El cluster " << id << " no existe\&." << endl;
103     else{
104         it->second\&.escriure();
105     }    
106 }
.fi
.SS "void Cjt_clusters::imprime_tabla_distancias () const"

.PP
Imprimeix la taula de distàncies entre clústers\&. 
.PP
\fBPrecondició\fP
.RS 4
Cert\&. 
.RE
.PP
\fBPostcondició\fP
.RS 4
S'ha imprès totes les distàncies entre clústers\&. 
.RE
.PP

.PP
Definició a la línia 109 del fitxer Cjt_clusters\&.cc\&.
.PP
.nf
109                                                   {
110   //Inv: És situa un iterador apuntant a la primera posició del
111   //mapa de distancies\&. Avançarà una posició fins que el punter apunti al final\&.
112   //A cada iteració s'impreix un identificador de espècie i s'executa un segon bucle\&.
113   for(map<string,map<string,double>>::const_iterator it = tabla_distancias_cluster\&.begin(); it != tabla_distancias_cluster\&.end(); ++it) {
114     cout << it->first << ":";
115       //Inv: És situa un iterador apuntant a la primera posició del
116     //map interior del map de distancies d'espècies\&. Avançarà una posició fins que el punter apunti al final\&.
117     //A cada iteració s'impreix un identificador d'espècie del map interior i la distancia respecte les dues espècies\&.
118     for (map<string,double>::const_iterator it2 = it->second\&.begin(); it2 != it->second\&.end(); ++it2) {
119         cout << " " << it2->first << " (" << it2->second << ")"; 
120     }
121     cout << endl;
122   }
123 }
.fi
.SS "void Cjt_clusters::imprime_arbol_filogenetico (\fBCjt_especies\fP & conjunt)"

.PP
Imprimeix l'arbre filogenètic dels clústers resultants\&. 
.PP
\fBPrecondició\fP
.RS 4
Cert\&. 
.RE
.PP
\fBPostcondició\fP
.RS 4
S'ha imprès l'arbre filogenètic de clústers pel canal standart de sortida\&. 
.RE
.PP

.PP
Definició a la línia 126 del fitxer Cjt_clusters\&.cc\&.
.PP
.nf
126                                                                     {
127     //el bool serveix per imprimir cada cop que executem el pas wpgma\&.
128     //fem un bucle fins que només quedi un element al mapa de clústers
129     bool fals = false;
130     inicialitza_clusters(conjunt,fals);
131 
132     if (map_clusters\&.size() != 0) {
133     while (map_clusters\&.size() > 1) {
134         ejecuta_paso_wpgma(fals);
135     }
136     map<string,Cluster>::const_iterator it = map_clusters\&.begin();
137     it->second\&.escriure(); //escriu el clúster resultant\&.
138     }
139     else cout << "ERROR: El conjunto de clusters es vacio\&." << endl;
140 
141 }
.fi
.SH "Documentació de les Dades Membre"
.PP 
.SS "map<string, \fBCluster\fP> Cjt_clusters::map_clusters\fC [private]\fP"

.PP
Estructura de dades on es guarden els clústers resultants\&. Conté un identificador i un clúster\&. 
.PP
Definició a la línia 42 del fitxer Cjt_clusters\&.hh\&.
.SS "map<string, map<string, double> > Cjt_clusters::tabla_distancias_cluster\fC [private]\fP"

.PP
Estructura de dades on guardem les distancies entre clústers\&. 
.PP
Definició a la línia 46 del fitxer Cjt_clusters\&.hh\&.

.SH "Autor"
.PP 
Generat automàticament per Doxygen per a Creació d'un arbre filogenètic\&. Xavier Coll Ribas a partir del codi font\&.


Cjt_clusters.cc
/** @file Cjt_clusters.cc
 * 	@brief Còdig de la classe Cjt_clusters.
*/

    //invariant El conjunt de clústers sempre estarà ordenat
	//alfabèticament i aquest no té mida màxima ni mínima. La taula
	//de distàncies entre clústers també està ordenada alfabèticament
	//amb els identificadors dels clústers.



#include "Cjt_clusters.hh"



// Constructores


Cjt_clusters::Cjt_clusters() {
    //Es crea el mapa de clústers i el de distancies buit.
    map_clusters = map<string,Cluster>();
    tabla_distancias_cluster = map<string,map<string,double>>();
}


// Consultores


double Cjt_clusters::dist_clust(const string& id, const string& id2) const{
    //Funció per consultar les distancies dins el mapa de 
    //distancies de clústers.
    		map<string,map<string,double>>::const_iterator it = tabla_distancias_cluster.find(id);
			map<string,double>::const_iterator it2 = it->second.find(id2);
			return it2->second;
}


pair<string,string> Cjt_clusters::min_dist() const {
    //Es busca els clústers a menor distancia i 
    //es retorna en una pair els dos clústers.
    double distancia = 101;
    pair<string,string> dists;
    for (map<string,map<string,double>>::const_iterator it = tabla_distancias_cluster.begin(); it != tabla_distancias_cluster.end(); ++it) {
        for (map<string,double>::const_iterator it2 = it->second.begin(); it2 != it->second.end(); ++it2) {
                if (it2->second < distancia ){
                distancia = it2->second;
                dists.first = it->first;
                dists.second = it2->first;
                }
            }

        }
    return dists;

}

// Modificadores


void Cjt_clusters::inicialitza_clusters(Cjt_especies& conjunt,const bool ve_de_func) {
    //Establim els iteradors del conjunt d'espècies per tal de poder fer el bucle
    //A cada iteració es modifica una string, agafant la id de una espècie i es
    //crea un clúster a partir de la string i la introdueix al mapa de clústers.
    map_clusters.clear();
    tabla_distancias_cluster.clear();
    conjunt.inicio(); // Iterador del conjunt d'espècies que apunta a la primera posició.
         while (not conjunt.final()) { //comprova que l'iterador no apunta al final
             string x;
             conjunt.actual(x); // La string x agafa el valor del iterador.
             Cluster clust(x);
             map_clusters.insert(make_pair(x, clust));
             conjunt.avanza(); //Avança posició del iterador.

        }
    //Es crea la taula de distancies a partir del conjunt.
    tabla_dist_clust(conjunt);
    if (ve_de_func) imprime_tabla_distancias();
}


void Cjt_clusters::ejecuta_paso_wpgma(const bool& a) {
    //S'executa el pas wpgma sempre i quan el tamany
    //del mapa de clústers sigui > 1.
    if (map_clusters.size() > 1) {
    pair<string,string> pair_dist = min_dist(); //pair amb els dos clústers a menor distancia.
    afegeix_especie_clusters(pair_dist); //Es fica els dos clústers i es fusionen en un nou clúster.
    elimina_especie_clusters(pair_dist.first); //S'elimina el clúster.
    elimina_especie_clusters(pair_dist.second); //S'elimina el clúster.
    if (a) imprime_tabla_distancias();
    }
    else cout << "ERROR: num_clusters <= 1" << endl;
}
 

//  Lectura i escriptura


void Cjt_clusters::imprime_cluster(string id) const {
    //Es posa un iterador apuntant al clúster indicat
    //Si no apunta al final, s'imprimeix, sino surt un missatge d'error.
    map<string,Cluster>::const_iterator it = map_clusters.find(id);
    if (it == map_clusters.end()) cout << "ERROR: El cluster " << id << " no existe." << endl;
    else{
        it->second.escriure();
    }    
}


void Cjt_clusters::imprime_tabla_distancias() const {
	//Inv: És situa un iterador apuntant a la primera posició del
	//mapa de distancies. Avançarà una posició fins que el punter apunti al final.
	//A cada iteració s'impreix un identificador de espècie i s'executa un segon bucle.
	for(map<string,map<string,double>>::const_iterator it = tabla_distancias_cluster.begin(); it != tabla_distancias_cluster.end(); ++it) {
		cout << it->first << ":";
    	//Inv: És situa un iterador apuntant a la primera posició del
		//map interior del map de distancies d'espècies. Avançarà una posició fins que el punter apunti al final.
		//A cada iteració s'impreix un identificador d'espècie del map interior i la distancia respecte les dues espècies.
		for (map<string,double>::const_iterator it2 = it->second.begin(); it2 != it->second.end(); ++it2) {
				cout << " " << it2->first << " (" << it2->second << ")"; 
		}
		cout << endl;
	}
}


void Cjt_clusters::imprime_arbol_filogenetico(Cjt_especies& conjunt)  {
    //el bool serveix per imprimir cada cop que executem el pas wpgma.
    //fem un bucle fins que només quedi un element al mapa de clústers
    bool fals = false;
    inicialitza_clusters(conjunt,fals);

    if (map_clusters.size() != 0) {
    while (map_clusters.size() > 1) {
        ejecuta_paso_wpgma(fals);
    }
    map<string,Cluster>::const_iterator it = map_clusters.begin();
    it->second.escriure(); //escriu el clúster resultant.
    }
    else cout << "ERROR: El conjunto de clusters es vacio." << endl;

}

// Mètodes privats


void Cjt_clusters::tabla_dist_clust(const Cjt_especies& conjunt) {
    //Consulta mitjançant el conjunt de espècies les distancies entre elles per 
    //passar-les al conjunt de clústers.
    for (map<string,Cluster>::const_iterator it = map_clusters.begin(); it != map_clusters.end(); ++it) {
        map<string, double> aux;
        for (map<string,Cluster>::const_iterator it2 = it; it2 != map_clusters.end(); ++it2) {
            if (it->first != it2->first) { //Comprova que el iterador sigui diferent al segon iterador.
            double x = conjunt.calcular_distancia(it->first, it2->first);
            aux.insert(make_pair(it2->first, x));
            }
        }
        tabla_distancias_cluster.insert(make_pair(it->first, aux));
    }
}


void Cjt_clusters::afegeix_especie_clusters(const pair<string,string>& dist) {
    //Aquí es modifica les distancies dels clústers fent servir
    //la formula de WPGMA.
    string fus = dist.first + dist.second;
    map<string,map<string,double>>::const_iterator it = tabla_distancias_cluster.begin();
    while (it != tabla_distancias_cluster.end() and it->first < fus) {
        tabla_distancias_cluster[it->first][fus] = (tabla_distancias_cluster[it->first][dist.first] + tabla_distancias_cluster[min(it->first,dist.second)][max(it->first,dist.second)])/2;
        ++it;
    }
    while (it != tabla_distancias_cluster.end()) {
        tabla_distancias_cluster[fus][it->first] = (tabla_distancias_cluster[dist.first][it->first] + tabla_distancias_cluster[min(it->first,dist.second)][max(it->first,dist.second)])/2;
        ++it;
    }
    map<string,Cluster>::const_iterator a = map_clusters.find(dist.first);
    map<string,Cluster>::const_iterator b = map_clusters.find(dist.second);
    Cluster c(a->second,b->second,dist_clust(dist.first,dist.second));
    map_clusters.insert(make_pair(fus,c));


}



void Cjt_clusters::elimina_especie_clusters(const string& id) {
    //Es posiciona un iterador apuntant a la posició de l'id demanat. El nom de l'iterador era per donar humor al còdig.
    map<string, Cluster>::const_iterator it_peix = map_clusters.find(id);
	map_clusters.erase(it_peix);
    //Es posiciona un iterador apuntant a la posició de l'id demanat. 
	map<string,map<string,double>>::const_iterator it = tabla_distancias_cluster.find(id);
    //Es posiciona un iterador apuntant a la primera posició del mapa de distancies.
	//Inv: Farà tantes iteracions fins que l'iterador sigui igual al primer que hem buscat. Així ens
	//estalviem bucles innecessaris. Aquest bucle elimina la espècie demanada del map interior.
	for (map<string,map<string,double>>::iterator it2 = tabla_distancias_cluster.begin(); it2 != it; ++it2) {
			it2->second.erase(it2->second.find(id));
		}
	//Eliminem la espècie del mapa gran de la taula de distancies.
	tabla_distancias_cluster.erase(it);

}



